driver/src/orderbook/streamed/state.rs

```diff
@@ -652,4 +691,93 @@ mod tests {
             2
         );
     }
+
+    #[test]
+    fn orderbook_batch_id() {
+        let mut state = state_with_fee();
+        let event = Deposit {
+            user: address(3),
+            token: address(0),
+            amount: 1.into(),
+            batch_id: 0,
+        };
+        state = state.apply_event(&Event::Deposit(event), 0).unwrap();
+        let balance = state.orderbook(Batch::Current).unwrap().0.next().unwrap().1;
+        assert_eq!(balance, U256::zero());
+        assert!(state.orderbook(Batch::Future(0)).is_err());
+        let balance = state
+            .orderbook(Batch::Future(1))
+            .unwrap()
+            .0
+            .next()
+            .unwrap()
+            .1;
+        assert_eq!(balance, U256::one());
+    }
+
+    #[test]
+    fn orderbook_partial_solution() {
+        let mut state = state_with_fee();
+        let event = TokenListing {
+            token: address(1),
+            id: 1,
+        };
+        for token in 0..2 {
+            let event = Deposit {
+                user: address(2),
+                token: address(token),
+                amount: 10.into(),
+                batch_id: 0,
+            };
+            state = state.apply_event(&Event::Deposit(event), 0).unwrap();
+        }
+        state = state.apply_event(&Event::TokenListing(event), 0).unwrap();
+        let event = OrderPlacement {
+            owner: address(2),
+            index: 0,
+            buy_token: 0,
+            sell_token: 1,
+            valid_from: 0,
+            valid_until: 10,
+            price_numerator: 5,
+            price_denominator: 5,
+        };
+        state = state.apply_event(&Event::OrderPlacement(event), 0).unwrap();
+        let event = Trade {
+            owner: address(2),
+            order_id: 0,
+            executed_sell_amount: 1,
+            executed_buy_amount: 2,
+            ..Default::default()
+        };
+        state = state.apply_event(&Event::Trade(event), 1).unwrap();
+
+        let balance = state
+            .orderbook(Batch::Current)
+            .unwrap()
+            .0
+            .collect::<HashMap<_, _>>();
+        assert_eq!(balance.get(&(address(2), 0)), Some(&10.into()));
+        assert_eq!(balance.get(&(address(2), 1)), Some(&10.into()));
+
+        // Error because the solution is not complete.
+        assert!(state.orderbook(Batch::Future(2)).is_err());
```

Comment:
Why is it not complete?

Reply:
Because we got Trade events but no SolutionSubmission event which means there could be more trades.

Reply:
Does the `SolutionSubmission` event contain any relevant information regarding how many trades we expect there to be?

Reply:
No